ledger/acctupdates.go

```diff
@@ -896,28 +926,140 @@ func (au *accountUpdates) initializeCaches(lastBalancesRound, lastestBlockRound,
 		}
 	}
 
-	for lastBalancesRound < lastestBlockRound {
-		next := lastBalancesRound + 1
+	skipAccountCacheMessage := make(chan struct{})
+	writeAccountCacheMessageCompleted := make(chan struct{})
+	defer func() {
+		close(skipAccountCacheMessage)
+		select {
+		case <-writeAccountCacheMessageCompleted:
+			au.log.Infof("initializeCaches completed initializing account data caches")
+		default:
+		}
+	}()
+	go func() {
+		select {
+		case <-time.After(initializingAccountCachesMessageTimeout):
+			au.log.Infof("initializeCaches is initializing account data caches")
+			close(writeAccountCacheMessageCompleted)
+		case <-skipAccountCacheMessage:
+		}
+	}()
 
-		blk, err = au.ledger.Block(next)
-		if err != nil {
-			return
+	blocksStream := make(chan bookkeeping.Block, initializeCachesReadaheadBlocksStream)
+	blockEvalFailed := make(chan struct{}, 1)
+	var blockRetrievalError error
+	go func() {
+		defer close(blocksStream)
+		for roundNumber := lastBalancesRound + 1; roundNumber <= lastestBlockRound; roundNumber++ {
+			blk, blockRetrievalError = au.ledger.Block(roundNumber)
+			if blockRetrievalError != nil {
+				return
+			}
+			select {
+			case blocksStream <- blk:
+			case <-blockEvalFailed:
+				return
+			}
+		}
+	}()
+
+	lastFlushedRound := lastBalancesRound
+	const accountsCacheLoadingMessageInterval = 5 * time.Second
+	lastProgressMessage := time.Now().Add(-accountsCacheLoadingMessageInterval / 2)
+
+	// rollbackSynchronousMode ensures that we switch to "fast writing mode" when we start flushing out rounds to disk, and that
+	// we exit this mode when we're done.
+	rollbackSynchronousMode := false
+	defer func() {
+		if rollbackSynchronousMode {
+			// restore default synchronous mode
+			au.dbs.Wdb.SetSynchronousMode(context.Background(), au.synchronousMode, au.synchronousMode >= db.SynchronousModeFull)
 		}
+	}()
 
+	for blk := range blocksStream {
```

Comment:
I probably miss something but how is reading from the `blocksStream` chan different from have a loop `for roundNumber := lastBalancesRound + 1; roundNumber <= lastestBlockRound; roundNumber++` ?